File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/BaseViewModel.java
package com.platdmit.simplecloudmanager.app.vm;

import androidx.lifecycle.ViewModel;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

public abstract class BaseViewModel extends ViewModel {
    protected final CompositeDisposable mCompositeDisposable = new CompositeDisposable();

    @Override
    protected void onCleared() {
        super.onCleared();
        mCompositeDisposable.clear();
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/vm/BaseViewModel.kt
package com.platdmit.simplecloudmanager.vm

import androidx.lifecycle.ViewModel
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.processors.BehaviorProcessor

abstract class BaseViewModel<T> : ViewModel() {
    protected val compositeDisposable = CompositeDisposable()
    protected val stateProvider: BehaviorProcessor<T> = BehaviorProcessor.create<T>()
    protected val messageProvider: BehaviorProcessor<String> = BehaviorProcessor.create<String>()

    override fun onCleared() {
        super.onCleared()
        messageProvider.onComplete()
        compositeDisposable.clear()
    }
}